libs/server/Custom/ExpandableMap.cs

```diff
@@ -189,31 +164,37 @@ public bool TryGetNextIdSafe(out int id)
         /// Try to update the actual size of the map based on the inserted item ID
         /// </summary>
         /// <param name="id">The inserted item ID</param>
-        /// <param name="noUpdate">True if should not do actual update</param>
         /// <returns>True if actual size should be updated (or was updated if noUpdate is false)</returns>
-        internal bool TryUpdateSize(int id, bool noUpdate = false)
+        private bool TryUpdateActualSize(int id)
         {
             var idx = id - minId;
 
             // Should not update the size if the index is out of bounds
             // or if index is smaller than the current actual size
             if (idx < 0 || idx < ActualSize || idx >= maxSize) return false;
 
-            if (!noUpdate)
-                ActualSize = idx + 1;
+            var oldActualSize = ActualSize;
+            var updatedActualSize = idx + 1;
+            while (oldActualSize < updatedActualSize)
+            {
+                var currActualSize = Interlocked.CompareExchange(ref actualSize, updatedActualSize, oldActualSize);
+                if (currActualSize == oldActualSize)
+                    break;
+                oldActualSize = currActualSize;
+            }
 
             return true;
         }
 
         /// <summary>
         /// Try to set item by ID
+        /// This method should only be called from a thread-safe context
         /// </summary>
         /// <param name="id">Item ID</param>
         /// <param name="value">Item value</param>
         /// <param name="noExpansion">True if should not attempt to expand the underlying array</param>
-        /// <param name="updateSize">True if should update actual size of the map</param>
         /// <returns>True if assignment succeeded</returns>
-        internal bool TrySetValue(int id, ref T value, bool noExpansion, bool updateSize)
+        private bool TrySetValueUnsafe(int id, ref T value, bool noExpansion)
```

Comment:
The invariant here should be that we are setting a entry that is currently null. We cannot _change_ an existing entry that is non-null. This is because in `TryGetFirstId`, there would be a race where we might apply the predicate on the old entry and succeed, but by the time we return that id, it is overwritten by a new entry. It might be better to debug assert that this invariant is true.